### ISSUE TITLE
pub(restricted) produces misleading diagnostic for private type in public interface lint

### ISSUE THREAD

**Issue by ubsan** [rust-lang/rust#33174]
``` rust
pub struct DoopGeneral<T> {
  pub(crate) var: T,
}
pub type Doop = DoopGeneral<Inner>;
pub(crate) struct Inner;
```

fails with `error: private type in public interface [E0446]`

https://play.rust-lang.org/?gist=35cee8425ac59ef27aa8898bd6343083&version=nightly&backtrace=0

**Comment by petrochenkov** [comment]
The error itself is correct, but the error messages are not updated yet.
`Doop` and `Inner` don't have the same scope (visibility), `Doop` is `pub` so its visibility is "all crates", `Inner` is `pub(crate)` so its visibility is "this crate".
Items with visibility `v` can't use items with visibilities `v' < v` in their interfaces.

**Comment by ubsan** [comment]
@petrochenkov Inner is not in the interface of Doop.

**Comment by estebank** [comment]
The current output has better wording:

```
error[E0446]: private type `Inner` in public interface
 --> src/main.rs:6:1
  |
6 | pub type Doop = DoopGeneral<Inner>;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type
```

It should probably be expanded to be

```
error[E0446]: private type `Inner` in public interface
 --> src/main.rs:6:1
  |
6 | pub type Doop = DoopGeneral<Inner>;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type
7 | pub(crate) struct Inner;
  | ----------------------- private type being leaked
```
